File: poc/player/evaluator.c
#include "evaluator.h"

#include "easing.h"

float mm_animation_value(const mm_animation_t *animation, float scene_time_ms)
{
    if(scene_time_ms <= animation->start_ms) return animation->from_value;
    if(animation->duration_ms <= 0.0f ||
       scene_time_ms >= animation->start_ms + animation->duration_ms) {
        return animation->to_value;
    }
    const float p = (scene_time_ms - animation->start_ms) / animation->duration_ms;
    const float eased = mm_ease(animation->easing, p);
    return animation->from_value + (animation->to_value - animation->from_value) * eased;
}

static void apply(mm_object_t *object, mm_property_t property, float value)
{
    switch(property) {
        case MM_PROP_OPACITY:         object->opacity = value; break;
        case MM_PROP_PROGRESS:        object->progress = value; break;
        case MM_PROP_VISIBLE:         object->visible = value != 0.0f; break;
        case MM_PROP_TRANSFORM_TX:    object->transform.tx = value; break;
        case MM_PROP_TRANSFORM_TY:    object->transform.ty = value; break;
        case MM_PROP_TRANSFORM_SCALE: object->transform.scale = value; break;
        case MM_PROP_DEFORM_AMPLITUDE:  object->deform.amplitude = value; break;
        case MM_PROP_DEFORM_WAVELENGTH: object->deform.wavelength = value; break;
        case MM_PROP_DEFORM_PHASE:      object->deform.phase = value; break;
        case MM_PROP_DEFORM_SWAY:       object->deform.sway = value; break;
        default: break;
    }
}

void mm_evaluate(mm_scene_t *scene, float scene_time_ms)
{
    /* Start from the authored values every time. The evaluator writes state in
     * place for the device's sake, so this reset is what keeps it a pure
     * function of the time: seeking backwards must give the same picture as
     * arriving forwards, and a property whose animation has not started yet
     * must show what the scene said, not what the last frame left. */
    for(int i = 0; i < scene->object_count; i++) {
        mm_object_t *object = &scene->objects[i];
        object->opacity = object->authored.opacity;
        object->progress = object->authored.progress;
        object->visible = object->authored.visible;
        object->transform = object->authored.transform;
        object->deform = object->authored.deform;
    }

    /* The animation that applies to a property is the last one, in declaration
     * order, whose start has been reached (decision 0006). One that has not
     * started must not impose its `from` value on an earlier phase, which is
     * what makes a phase list -- build up, then decay -- work at all. */
    for(int i = 0; i < scene->animation_count; i++) {
        const mm_animation_t *animation = &scene->animations[i];
        if(animation->target < 0 || animation->target >= scene->object_count) continue;
        if(animation->start_ms > scene_time_ms) continue;

        bool superseded = false;
        for(int j = i + 1; j < scene->animation_count; j++) {
            const mm_animation_t *later = &scene->animations[j];
            if(later->target == animation->target &&
               later->property == animation->property &&
               later->start_ms <= scene_time_ms) {
                superseded = true;
                break;
            }
        }
        if(superseded) continue;

        apply(&scene->objects[animation->target], animation->property,
              mm_animation_value(animation, scene_time_ms));
    }
}

File: mementum_node/core/evaluator.py
"""The evaluator: ``state = scene.evaluate(sceneTime)`` (proposal §10, §19).

Pure by construction. There is no accumulator, no frame counter and no
dependence on call order:

    sceneTime = sharedNow() - displayAt
    state     = evaluate(scene, sceneTime)

Seeking to an arbitrary time therefore gives exactly the result of arriving
there by stepping -- which is what makes a dropped frame, a missed PLAY and a
late join the same problem with one answer.

Hold semantics: before an animation starts the property holds ``from``; after it
ends it holds ``to``.

Where several animations drive the same property, the one that applies is the
last, in declaration order, whose ``start`` has been reached -- so a phase list
(build up, then decay) works, and a property with no animation yet started
keeps its authored value (decision 0006). Blending overlapping animations is
deliberately not a thing in v1.
"""

from __future__ import annotations

from dataclasses import replace

from .easing import ease
from .scene import Animation, Layer, Scene, SceneObject

__all__ = ["evaluate", "value_at"]


def value_at(anim: Animation, scene_time: float) -> float:
    """The animated value at ``scene_time`` in milliseconds. Pure."""
    if scene_time <= anim.start:
        return anim.from_value
    if anim.duration <= 0 or scene_time >= anim.end:
        return anim.to_value
    p = (scene_time - anim.start) / anim.duration
    return anim.from_value + (anim.to_value - anim.from_value) * ease(anim.easing, p)


def evaluate(scene: Scene, scene_time: float) -> Scene:
    """Return a copy of ``scene`` with every animated property resolved at
    ``scene_time`` (milliseconds since ``displayAt``). Pure: the same scene and
    the same time always give the same result, and the input is not mutated."""
    if not scene.animations:
        return scene

    # Pick the governing animation per (object, property) before evaluating
    # any of them: the last one to have started wins, and one that has not
    # started yet must not impose its ``from`` value on earlier phases.
    governing: dict[str, dict[str, Animation]] = {}
    for anim in scene.animations:
        if anim.start > scene_time:
            continue
        governing.setdefault(anim.target, {})[anim.property] = anim

    resolved: dict[str, dict[str, float]] = {
        target: {prop: value_at(anim, scene_time) for prop, anim in props.items()}
        for target, props in governing.items()
    }

    layers: list[Layer] = []
    for layer in scene.layers:
        objects: list[SceneObject] = []
        for obj in layer.objects:
            props = resolved.get(obj.id)
            if props:
                for name, value in props.items():
                    obj = obj.with_property(name, value)
            objects.append(obj)
        layers.append(replace(layer, objects=tuple(objects)))
    return replace(scene, layers=tuple(layers))
